sky/templates/kubernetes-port-forward-proxy-command.sh.j2

```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+set -uo pipefail
+
+# Checks if socat is installed
+if [ ! "$(which socat)" ]; then
+  echo "Using 'port-forward' mode to run ssh session on Kubernetes instances requires 'socat' to be installed. Please install 'socat'" >&2
+  exit
+fi
+
+# If the port-forward command is already running on the jump pod, kill it
+pgrep -f "kubectl port-forward svc/{{ ssh_jump_name }} {{ local_port }}:22" > /dev/null
+if [ $? -eq 0 ]; then
+    pkill -f "kubectl port-forward svc/{{ ssh_jump_name }} {{ local_port }}:22"
+fi
```

Comment:
I ran into a bug where creating multiple ssh connections in parallel (e.g., multiple ssh windows or parallel sky launches) would allow only one connection to live at a time. I.e., the previous connection gets abruptly terminated.

To replicate:
1. `sky launch -c test`
2. Open a terminal and run `ssh test`
3. While the previous terminal is running, open a new terminal and run `ssh test`
4. The ssh connection created in step 2 gets broken.

```
(base) sky@test-ray-head:~$ /Users/romilb/.sky/port-forward-proxy-cmd.sh: line 30: 32784 Terminated: 15          kubectl port-forward svc/sky-sshjump-2ea485ef 23100:22
                                           /Users/romilb/.sky/port-forward-proxy-cmd.sh: line 31: kill: (32784) - No such process
     client_loop: send disconnect: Broken pipe
client_loop: send disconnect: Broken pipe
``` 

We may need to 
a) find a better way to handle open `kubectl port-forward` instead of killing them. Perhaps check if it is in use or not
b) If the port is in use, use another local_port to create the new connection

Reply:
@landscapepainter - here's a script that works and handles multiple SSH connections:

```
#!/usr/bin/env bash
set -uo pipefail

# Checks if socat is installed
if ! command -v socat > /dev/null; then
  echo "Using 'port-forward' mode to run ssh session on Kubernetes instances requires 'socat' to be installed. Please install 'socat'" >&2
  exit
fi

# Checks if lsof is installed
if ! command -v lsof > /dev/null; then
  echo "Checking port availability requires 'lsof' to be installed. Please install 'lsof'" >&2
  exit 1
fi

# Function to check if port is in use
is_port_in_use() {
    local port="$1"
    lsof -i :${port} > /dev/null 2>&1
}

# Start from a fixed local port and increment if in use
local_port={{ local_port }}
while is_port_in_use "${local_port}"; do
    local_port=$((local_port + 1))
done

# Establishes connection between local port and the ssh jump pod
kubectl port-forward svc/{{ ssh_jump_name }} "${local_port}":22 &

# Terminate the port-forward process when this script exits.
K8S_PORT_FWD_PID=$!
trap "kill $K8S_PORT_FWD_PID" EXIT

# checks if a connection to local_port of 127.0.0.1:[local_port] is established
while ! nc -z 127.0.0.1 "${local_port}"; do
    sleep 0.1
done

# Establishes two directional byte streams to handle stdin/stdout between
# terminal and the jump pod.
# socat process terminates when port-forward terminates.
socat - tcp:127.0.0.1:"${local_port}"
```

Reply:
Updated the script to the version above and added a check for `lsof` installation in `authentication.py`. Reason why we need to check for installations in both `authentication.py` and `port-fordward-proxy-cmd.sh`: https://github.com/skypilot-org/skypilot/pull/2412#discussion_r1299460245

Reply:
@romilbhardwaj Do you think it's necessary to set an upper bound on possible `local_port` usage? It can limit the number of concurrent ssh sessions, but was wondering if we should set a range on available `local_port` for this purpose.